### PR TITLE
Rework AFSelector to normalize and simplify its locking

### PR DESCRIPTION
**Pull request by ThePumpingLemma**
Note that this PR is split into three commits; the first two commits are mainly renames to make the implementation more consistent with the Java standard library's implementations. For reviewing, strongly recommend ignoring whitespace changes.

Changes:
* Place bulk of `select0` operations under `this` and
  `publicSelectedKeys` locks to reduce locking/unlocking thrash given
  frequency of locking operations
* Lock `publicSelectedKeys` instead of `selectedKeys` since that is what
  clients of AFSelector are locking on
* Eschew uses of synchronized methods in favor of locking `this` via
  synchronized blocks
* Use assertions to check lock invariants

### CODE REVIEW

File: junixsocket-common/src/main/java/org/newsclub/net/unix/AFSelector.java [hunk 1]
```diff
@@ -110,47 +111,51 @@ public int select() throws IOException {
...
-        consumeAllBytesAfterPoll();
-        setOpsReady(pfd); // updates keysSelected and numKeysSelected
-      }
-      return selectedKeysSet.size();
     }
   }
 
-  private synchronized void consumeAllBytesAfterPoll() throws IOException {
+  private void consumeAllBytesAfterPoll() throws IOException {
+    assert Thread.holdsLock(this);
```
Comment (kohlschuetter) [inline junixsocket-common/src/main/java/org/newsclub/net/unix/AFSelector.java:157]: Shouldn't this be assert Thread.holdsLock(publicSelectedKeys)) ?

File: junixsocket-common/src/main/java/org/newsclub/net/unix/AFSelector.java [hunk 2]
```diff
@@ -110,47 +111,51 @@ public int select() throws IOException {
 
   @SuppressWarnings("PMD.CognitiveComplexity")
   private int select0(int timeout) throws IOException {
-    PollFd pfd;
     synchronized (this) {
```
Comment (kohlschuetter) [inline junixsocket-common/src/main/java/org/newsclub/net/unix/AFSelector.java:114]: Why lock on `this` when we now also lock on `publicSelectedKeys`?

**Comment by kohlschuetter** [comment]
There's a lot going on in AFSelector for sure, and it's not the most readable code. I'm not sure that this PR resolves this issue.

As a general rule, try simplifying your change so it's easy to follow what is actually changed. Please do not rename methods/fields just because, it really makes things hard to follow, prevents unrelated changes from merging cleanly, and honestly in this case, it is a matter of consistency and maybe taste, so I don't think we need these particular changes.

I can't remember exactly why, but the synchronized-locking was split around the JNI poll method for a reason. That reason may have been wrong in the first place, or things have changed, but I would want to make sure we don't introduce yet another regression.

What if we focus the change on locking on `publicSelectedKeys` instead of `selectedKeys`?

File: junixsocket-common/src/main/java/org/newsclub/net/unix/AFSelector.java (hunk 2 above)
Comment (ThePumpingLemma) [reply junixsocket-common/src/main/java/org/newsclub/net/unix/AFSelector.java:114]: That's a good question. In the standard lib, they lock on both for most of the implementation, but I think we can push down the lock for selected keys to leave `initPollFd` only locked under `this`.

File: junixsocket-common/src/main/java/org/newsclub/net/unix/AFSelector.java (hunk 1 above)
Comment (ThePumpingLemma) [reply junixsocket-common/src/main/java/org/newsclub/net/unix/AFSelector.java:157]: This case is interesting. It doesn't actually need to lock on `publicSelectedKeys` for correctness, but then we would have to lock, drop, then immediately lock on `publicSelectedKeys` again if we didn't want to consume bytes under the lock. I think we can reorder some of this to reduce what is under the lock.

**Comment by ThePumpingLemma** [comment]
> There's a lot going on in AFSelector for sure, and it's not the most readable code. I'm not sure that this PR resolves this issue, though.

This feedback is fair.

> As a general rule, try simplifying your change so it's easy to follow what is actually changed. Please do not rename methods/fields just because, it really makes things hard to follow, prevents unrelated changes from merging cleanly, and honestly in this case, it is a matter of consistency and maybe taste, so I don't think we need these particular changes.

Okay, I am working on a separate PR with only the locking changes.

> I can't remember exactly why, but the synchronized-locking was split around the JNI poll method for a reason. That reason may have been wrong in the first place, or things have changed, but I would want to make sure we don't introduce yet another regression.

This part is odd to me as well. In [Java's standard library](https://github.com/openjdk/jdk/blob/7c7f8ea30da7fe552bcd4f2b593fa9aad27dcdb4/src/java.base/share/classes/sun/nio/ch/SelectorImpl.java#L120-L136), calls to native implementations like epoll and kqueue don't drop the locks and, as far as I can tell, it doesn't negatively impact junixsocket either. One thing I did notice is that [Java's variants omit the `begin`/`end` synchronization when invoking `selectNow`](https://github.com/openjdk/jdk/blob/7c7f8ea30da7fe552bcd4f2b593fa9aad27dcdb4/src/java.base/share/classes/sun/nio/ch/SelectorImpl.java#L96-L108).

> What if we focus the change on locking on `publicSelectedKeys` instead of `selectedKeys`?

On it!

**Comment by ThePumpingLemma** [comment]
Okay, I took another pass and this PR only adjusts locking: https://github.com/kohlschutter/junixsocket/pull/147

I think it also addresses your comments with locking on `this` versus `selectedKeysPublic` by making what is under the locks meaningfully different.

**Comment by kohlschuetter** [comment]
Continuing discussion in #147.
